docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+
+We want to investigate what *High Availability* - HA - means for a project like
+the s3gw.
+
+If we identify the meaning of HA as the ability to have N independent pipelines
+so that the failure of some of these does not affect the user's operations,
+this is something that could be not easy to achieve with the s3gw.
+
+With pipeline we mean all the chain from the ingress to the `PV`:
+
+```mermaid
+flowchart LR
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+```
+
+The difficulty with a project like the s3gw is due to the fact that basically
+we have one process: the `radosgw`, that has the *exclusive* access to one resource:
+the Kubernetes `PV` where the S3 buckets and objects are persisted.
+
+Speaking in theory, 3 HA models are possible with s3gw:
+
+1. **Active-Active** pipelines
+2. **Active-Passive** pipelines
+3. **One Active** pipeline (no Passive pipelines preloaded)
+
+## Active-Active pipelines
+
+This is the model that guarantees the best HA characteristics.
+The *Active-Active* model must implement *true independent pipelines*
+where all the pieces are replicated.
+
+The immediate consequence of this statement is that every pipeline being part of
+the same logical s3gw service must bind to a different `PV`; one per `radosgw` process.
+All the `PV`s for the same logical s3gw, must therefore, sync their data.
+
+This implies that with this model, the s3gw must implement an efficient replication
+model across all the pipelines being part of the same logical service.
+
+Under these conditions, a client accessing the S3 service would experience a failure
+only for *the active operation* that were happening on the failed s3gw pipeline.
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+    ingress ==> radosgw2(radosgw) ==> PV2((PV))
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: red
+    linkStyle 2 stroke: red
+```
+
+The subsequent access to the S3 service would result in a successful operation
+because the load balancer would immediately reroute to another active pipeline.
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+    ingress ==> radosgw2(radosgw) ==> PV2((PV))
+
+    linkStyle 0 stroke: green
+    linkStyle 1 stroke: red
+    linkStyle 2 stroke: red
+    linkStyle 3 stroke: green
+    linkStyle 4 stroke: green
+```
+
+This model is potentially hugely expensive for the s3gw project
+because there isn't an implementation for the data replication with the SFS backend.
+
+## Active-Passive pipelines
+
+This model assumes N pipelines to be "allocated" on the cluster at the same time,
+but only one, the *active* pipeline, owns the *exclusive ownership*
+over the shared `PV`.
+
+Should the active pipeline suffer a failure, the next elected active pipeline,
+chosen to replace the old one, should pay a *time window* necessary to "transfer"
+the ownership over the shared `PV`.
```

Comment:
As explained above, there's also a latency to active/active setups to detect and recover from faults.

The active plus warm standby operation here has no meaningful advantages over the passive/fail-over option; loading the s3gw pod is the cheapest part of the whole process, compared to fault detection (usually a timeout), mounting (journal recovery) of the file system, the process running the SQLite etc recovery on start, etc.

We'd pay for this with complexity (and resource consumption while in standby) that likely would only give us very marginal benefits at best. I'd argue that beyond explaining why, we do not need to further investigate this option.

Reply:
roger :+1: